pandas/core/groupby/generic.py

```diff
@@ -622,33 +625,27 @@ def value_counts(
         ids, val = ids[mask], val[mask]
 
         if bins is None:
-            lab, lev = algorithms.factorize(val, sort=True)
+            lab, lev = cast(
+                Tuple[np.ndarray, Index], algorithms.factorize(val, sort=True)
```

Comment:
If possible, it would be great to instead add overloads to [factorize](https://github.com/pandas-dev/pandas/blob/1b3bb88008011f8ce7e3ec65fcc0c31fb838352b/pandas/core/algorithms.py#L576).

Reply:
the revealed type of `val` is `Any`. `val = self.obj._values`, The revealed type of `self.obj` is correctly `Series`, but the `_value` property of Series is untyped.

So I don't think overloading factorize would help here till `val` does not resolve to `Any`

I don't think in general should replace ignores with casts. We should be careful that we are not lying to mypy. If it is clear that the return will definitely be `Index`, could maybe add a code comment adjacent to to cast with the justification of the cast.

If doing that I would only add the cast for `lev` and leave `lab` to be inferred. But I would probably prefer to leave the ignore until the other issues are resolved.

(I think we still have much of the lower level code to type)

Reply:
Based on the signature of `algorithms.factorize` I would expect that `lev` is assumed to be `np.ndarray | Index`. I would assume that there is a relationship between the first argument and the type of the second return value (`lev`) but i'm not sure how easy it will be to write overloads for this.
https://github.com/pandas-dev/pandas/blob/1b3bb88008011f8ce7e3ec65fcc0c31fb838352b/pandas/core/algorithms.py#L581

Reply:
yep, wherever we have union return types we probably need overloads eventually, but my point is that I don't think that adding the overloads alone would fix the issue, since if the argument is `Any`, we would still return the union.

(side note: I'm sure i've tackled factorize in one of my stashes in the past, but the problem in many cases like this is overlapping overloads. (in this case IIRC it's the categorical) We would need to add ignores (which is normally difficult to get through the review process) and to be sure that the overloads are actually working as intended by matching the correct overload need tests in place to test the overloads. This has been in the pipeline for quite a while and personally I have been trying to avoid adding overloads till the test framework is in place)

Reply:
Judging by the discussion it seems that it is better to leave as is for now. I will remove it from the PR